### DISCUSSION TITLE
Completely lost with installing core runtime (dotnet 8)

### DISCUSSION THREAD

**Discussion by Lindsay-Mathieson** [oleg-shilo/wixsharp#1639] [Q&A] [answered]
Have the main installer sorted - does my desktop app and service

Want to have it optionally install the dotnet 8 desktop & asp.net runtime and for the life of me can't figure out where to start, I know it involves something to do with "bootstrapping" but all the docs/examples I find online are old, incomplete and seem to assume a lot of prior knowledge or are so massively complicated, installing exercising every feature that I can't decipher what is happening.

Can anyone point me to a simple working sample of say installing a dotnet 7 or 8 runtime before the main installer?

This simple command line does all that I want:
```cmd
winget install --silent --accept-source-agreements --accept-package-agreements dotnet-desktop-8 aspnetcore-8
```

But I can't invoke it from the installer as that would nest the msiexec process.

Thanks.

**Comment by NJullienSweet** [comment, marked as answer]
I'm also installing .NET 8 on my side. I'm using a bootstrapper indeed.

To do so, I'm creating an `ExePackage` and specifying the URL where the user machine will be able to download the installer. You have also have to supply the installer's hash and its size to make sure that the machine has the one you intended to install.

```cs
const string DotnetRuntimeDownloadUrl = "https://download.visualstudio.microsoft.com/download/pr/b280d97f-25a9-4ab7-8a12-8291aa3af117/a37ed0e68f51fcd973e9f6cb4f40b1a7/windowsdesktop-runtime-8.0.0-win-x64.exe";
(var hash, var size) = await GetHashAndSizeAsync(DotnetRuntimeDownloadUrl);

var dotnetInstaller = new ExePackage
{
	Id = "DotNet8",
	InstallArguments = "/install /quiet /norestart",
	RepairArguments = "/silent /repair /norestart",
	UninstallArguments = "/silent /uninstall /norestart",
	PerMachine = true,
	DetectCondition = "DOTNET_RUNTIME_CHECK >= v8",
	Payloads = new Payload[] {
		new ExePackagePayload
		{
			Name = "windowsdesktop-runtime-8.0.0-win-x64.exe",
			DownloadUrl = DotnetRuntimeDownloadUrl,
			Hash = hash,
			Size = (int)size.Bytes
		}
	}
};

// specifying dotnetInstaller then yourMsiHere will make sure to install these in this exact order.
var bundle = new Bundle("BundleName", dotnetInstaller, yourMsiHere);

// Rest of your WixSharp setup
```

The DetectCondition (DOTNET_RUNTIME_CHECK >= v8) results of this:

```cs
// Adds .NET compatibility check to the bundle.
bundle.AddXml("Wix", new DotNetCompatibilityCheck(
	"WindowsDesktopRuntimeCheck",
	"DOTNET_RUNTIME_CHECK",
	DotNetCompatibilityCheck.RollForwardPolicy.LatestMinor,
	DotNetCompatibilityCheck.RuntimeType.Desktop,
	Platform.x64,
	new Version(8, 0, 0, 0)).ToXmlFragment());
```

**Reply by Lindsay-Mathieson** [reply]
Thanks!

This line
```c#
var bundle = new Bundle("BundleName", dotnetInstaller, yourMsiHere);
```
Thats where you integrate the actual applications installer? So we have two install projects?
* MSI App Installer
* EXE runtime installer that includes and runs the MSI Installer?

**Reply by NJullienSweet** [reply]
The 2nd to nth parameters on new Bundle(...) are chain elements. That means that the Bundle EXE will install all of these in the order you supplied them.
Your generally want to install first your dependencies (i.e. .NET, Visual C++ Redistribuable, etc) and then your actual app MSI.
Building this bundle will output a single EXE with your two (or more) installers embedded in. The bundle EXE will take care of installing all of this in the correct order.

**Reply by Lindsay-Mathieson** [reply]
> The 2nd to nth parameters on new Bundle(...) are chain elements. That means that the Bundle EXE will install all of these in the order you supplied them.

Thanks!

**Reply by oleg-shilo** [reply]
@NJullienSweet, kudos on the excellent use of  `AddXml` extension. 👍 

Maybe you can share the code for `DotNetCompatibilityCheck` class? To make your shared code complete.

**Reply by NJullienSweet** [reply]
Indeed, I forgot I made it myself 😁

```cs
// DotNetCompatibilityCheck.cs

using System;
using System.Xml.Linq;
using WixSharp;

/// <summary>
/// Represents a class for checking .NET compatibility.
/// </summary>
public class DotNetCompatibilityCheck
{
	/// <summary>
	/// Specifies the roll forward policy for .NET compatibility check.
	/// </summary>
	public enum RollForwardPolicy
	{
		Disable,
		LatestPatch,
		LatestMinor,
		LatestMajor,
		Minor,
		Major
	}

	/// <summary>
	/// Specifies the runtime type for .NET compatibility check.
	/// </summary>
	public enum RuntimeType
	{
		Desktop,
		AspNet,
		Core,
	}

	private const string NetFxSchemaUrl = "http://wixtoolset.org/schemas/v4/wxs/netfx";

	/// <summary>
	/// Gets the identifier of the compatibility check.
	/// </summary>
	public readonly string Id;

	/// <summary>
	/// Gets the name of the property to set.
	/// </summary>
	public readonly string PropertyName;

	/// <summary>
	/// Gets the roll forward policy for the compatibility check.
	/// </summary>
	public readonly RollForwardPolicy RollForward;

	/// <summary>
	/// Gets the runtime type for the compatibility check.
	/// </summary>
	public readonly RuntimeType Runtime;

	/// <summary>
	/// Gets the platform for the compatibility check.
	/// </summary>
	public readonly Platform Platform;

	/// <summary>
	/// Gets the version for the compatibility check.
	/// </summary>
	public readonly Version Version;

	/// <summary>
	/// Initializes a new instance of the <see cref="DotNetCompatibilityCheck"/> class.
	/// </summary>
	/// <param name="id">The identifier of the compatibility check.</param>
	/// <param name="propertyName">The name of the property to set.</param>
	/// <param name="rollForward">The roll forward policy.</param>
	/// <param name="runtime">The runtime type.</param>
	/// <param name="platform">The platform.</param>
	/// <param name="version">The version.</param>
	public DotNetCompatibilityCheck(string id, string propertyName, RollForwardPolicy rollForward, RuntimeType runtime, Platform platform, Version version)
	{
		Id = id;
		PropertyName = propertyName;
		RollForward = rollForward;
		Runtime = runtime;
		Platform = platform;
		Version = version;
	}

	/// <summary>
	/// Converts the compatibility check to an XML fragment.
	/// </summary>
	/// <returns>The XML fragment representing the compatibility check.</returns>
	public string ToXmlFragment()
	{
		XNamespace netfx = NetFxSchemaUrl;
		var element = new XElement(netfx + "DotNetCompatibilityCheck",
				new XAttribute("Id", Id),
				new XAttribute("Property", PropertyName),
				new XAttribute("RollForward", RollForward.GetStringValue()),
				new XAttribute("RuntimeType", Runtime.GetStringValue()),
				new XAttribute("Platform", Platform),
				new XAttribute("Version", Version.ToString(3)));

		var fragment = new XElement("Fragment", element);
		return fragment.ToString();
	}

}

/// <summary>
/// Provides extension methods for the <see cref="DotNetCompatibilityCheck.RollForwardPolicy"/> enum.
/// </summary>
public static class RollForwardPolicyExtensions
{
	/// <summary>
	/// Gets the string value of the roll forward policy.
	/// </summary>
	/// <param name="policy">The roll forward policy.</param>
	/// <returns>The string value of the roll forward policy.</returns>
	public static string GetStringValue(this DotNetCompatibilityCheck.RollForwardPolicy policy)
	{
		switch (policy)
		{
			case DotNetCompatibilityCheck.RollForwardPolicy.Disable:
				return "disable";
			case DotNetCompatibilityCheck.RollForwardPolicy.LatestPatch:
				return "latestPatch";
			case DotNetCompatibilityCheck.RollForwardPolicy.LatestMinor:
				return "latestMinor";
			case DotNetCompatibilityCheck.RollForwardPolicy.LatestMajor:
				return "latestMajor";
			case DotNetCompatibilityCheck.RollForwardPolicy.Minor:
				return "minor";
			case DotNetCompatibilityCheck.RollForwardPolicy.Major:
				return "major";
			default:
				return string.Empty;
		}
	}
}

/// <summary>
/// Provides extension methods for the <see cref="DotNetCompatibilityCheck.RuntimeType"/> enum.
/// </summary>
public static class RuntimeTypeExtensions
{
	/// <summary>
	/// Gets the string value of the runtime type.
	/// </summary>
	/// <param name="runtimeType">The runtime type.</param>
	/// <returns>The string value of the runtime type.</returns>
	public static string GetStringValue(this DotNetCompatibilityCheck.RuntimeType runtimeType)
	{
		switch (runtimeType)
		{
			case DotNetCompatibilityCheck.RuntimeType.Desktop:
				return "desktop";
			case DotNetCompatibilityCheck.RuntimeType.AspNet:
				return "aspnet";
			case DotNetCompatibilityCheck.RuntimeType.Core:
				return "core";
			default:
				return string.Empty;
		}
	}
}

**Reply by oleg-shilo** [reply]
Good stuff.

You can further simplify it a little by utilizing WixSharp XML serialization:
You can also use ready-to-go `WixSharp.RoolForward` instead of `DotNetCompatibilityCheck.RollForwardPolicy`. It has the all enum values properly named so you don't even need to implement `GetStringValue`. The `ToXElement` will do it all for you.
The same comes for `WixSharp.RuntimeType`.

I just tested it and it seems like you can achieve the desired outcome with this implementations too:

```C#
        public class DotNetCompatibilityCheck : WixEntity
        {
            [Xml]
            public new string Id { get => base.Id; set => base.Id = value; }

            [Xml]
            public readonly string PropertyName;

            [Xml]
            public readonly RollForward RollForward;

            [Xml]
            public readonly RuntimeType Runtime;

            [Xml]
            public readonly Platform Platform;

            [Xml]
            public readonly Version Version;

            public DotNetCompatibilityCheck(string id, string propertyName, RollForward rollForward, RuntimeType runtime, Platform platform, Version version)
            {
                Id = id;
                PropertyName = propertyName;
                RollForward = rollForward;
                Runtime = runtime;
                Platform = platform;
                Version = version;
            }
        }

        [Fact]
        public void MapObjectToToXml()
        {
            var entity = new DotNetCompatibilityCheck(
                "WindowsDesktopRuntimeCheck",
                "DOTNET_RUNTIME_CHECK",
                RollForward.latestMinor,
                RuntimeType.desktop,
                Platform.x64,
                new Version(8, 0, 0, 0));

            var name = WixExtension.NetFx.ToXName("DotNetCompatibilityCheck");

            var xml = entity.ToXElement(name).ToString();
        }

```

**Reply by oleg-shilo** [reply]
In fact I will probably add it to the codebase. If you are OK with me using your contribution.

**Reply by oleg-shilo** [reply]
@NJullienSweet, I just realized we don't need to convert your class to the string. You can add it as an XElement object. 

but the best way to do that is to do this with a special IXmlAware interface. The code is cleaner then:

```c#
public class DotNetCompatibilityCheck : WixEntity, IXmlAware
{
    . . .
    public XElement ToXml()
        => this.ToXElement(WixExtension.NetFx.ToXName("DotNetCompatibilityCheck"));
}

...

bundle.AddWixFragment("Wix", new DotNetCompatibilityCheck(
	"WindowsDesktopRuntimeCheck",
	"DOTNET_RUNTIME_CHECK",
	DotNetCompatibilityCheck.RollForwardPolicy.LatestMinor,
	DotNetCompatibilityCheck.RuntimeType.Desktop,
	Platform.x64,
	new Version(8, 0, 0, 0)));
```

**Reply by NJullienSweet** [reply]
When I try to make it a WixEntity I'm ending up with errors about a generated `DotNetCompatibilityCheckRef` and I don't really have enough time to fix it today, so I'll wait for your own official version! 😁
Glad I could help though!

**Reply by oleg-shilo** [reply]
Not a prob. Will do.
`DotNetCompatibilityCheckRef` might be caused by not having `bundle.Include(WixExtension.NetFx)`, but I will see.

**Reply by oleg-shilo** [reply]
No, I was wrong. The problem is caused by WixExtension.NetFx not feigning DotNetCompatibilityCheckRef element which seems to be the normal way of integrating Fragment elements. IE as for `ProductSearch`:

![image](https://github.com/user-attachments/assets/db539348-d980-460b-84cf-bacfb1c8a6e4)

however in case of DotNetCompatibilityCheckRef it throws the error:

![image](https://github.com/user-attachments/assets/9456ab7b-53c7-4872-aee9-8f7b684d13f9)

Thus if one wants to use your approach but with the very current WixSharp the working code is:

```C#
bundle.AddXml("Wix", new XElement("Fragment", new DotNetCompatibilityCheck(
    "WindowsDesktopRuntimeCheck",
    "DOTNET_RUNTIME_CHECK",
    RollForward.latestMinor,
    RuntimeType.desktop,
    Platform.x64,
    new Version(8, 0, 0, 0)).ToXElement(WixExtension.NetFx.ToXName("DotNetCompatibilityCheck"))).ToString());
```

But of course, I am going to allow a more natural approach:
```c#
bundle.AddWixFragment("Wix", new DotNetCompatibilityCheck(
    "WindowsDesktopRuntimeCheck",
    "DOTNET_RUNTIME_CHECK",
    RollForward.latestMinor,
    RuntimeType.desktop,
    Platform.x64,
    new Version(8, 0, 0, 0)));
```

**Reply by oleg-shilo** [reply]
Done. 
Release v2.4.1

-----

There are now various ways to add the DotNetCompatibilityCheck to the bundle as XML but this is the recommended one:

```c#
var check = new DotNetCompatibilityCheck(
    "DOTNET_RUNTIME_CHECK",
    RollForward.latestMinor,
    RuntimeType.desktop,
    Platform.x64,
    new Version(8, 0, 0, 0));

bundle.GenericItems.Add(check);
```

**Reply by Lindsay-Mathieson** [reply]
> ```cs
> (var hash, var size) = await GetHashAndSizeAsync(DotnetRuntimeDownloadUrl);
> ```

Do you have the src for this function? Does it read the hash and size from the page directly?

Currently I'm implementing it by actually downloading the file can calculating them, but that seems sub optimal.

**Comment by Lindsay-Mathieson** [comment]
Fantastic stuff, carving out some time this weekend to test it.

Related question - when installing dotnet runtimes as dependences of our apps, should we really be uninstalling them as part of our apps uninstaller?

**Reply by NJullienSweet** [reply]
Most of the software out there doesn't. It has the downside of filling PCs with dependencies that could stay unused in a corner.
In a perfect world, Windows would provide an API to know if other software uses it 😁
In our imperfect world, in my opinion a better approach would be to ask the user whether to uninstall it or no, so it becomes their responsibility 😂

**Reply by Lindsay-Mathieson** [reply]
Is there a way for a bundle to invoke an command on the pc, rather executing an included or remote installer?

I'd like to just run

```cmd
winget install --silent --accept-source-agreements --accept-package-agreements dotnet-desktop-8 aspnetcore-8
```
Which installs the latest dot 8 components, rather than download and execute a specific dotnet 8 version

**Reply by Lindsay-Mathieson** [reply]
Have a working bootstrap exe, that installs dotnet desktop and asp runtimes - thanks very much

It also installs my MSI, but none of my managed dialogs are showing, just the generic license screen, which I had disabled.

My MSI Package

```c#
            var msiPackage = new MsiPackage(builtProject)
            {
                DisplayInternalUI = true,
                EnableFeatureSelection = true,
            };
```

and the setup I have
```c#
            project.ManagedUI = new ManagedUI();

            project.ManagedUI.InstallDialogs.Add(Dialogs.Welcome)
                                            //.Add(Dialogs.Licence)
                                            //.Add(Dialogs.SetupType)
                                            .Add(Dialogs.Features)
                                            .Add(Dialogs.InstallDir)
                                            .Add(Dialogs.Progress)
                                            .Add(Dialogs.Exit);

            project.ManagedUI.ModifyDialogs.Add(Dialogs.MaintenanceType)
                                           .Add(Dialogs.Features)
                                           .Add(Dialogs.Progress)
                                           .Add(Dialogs.Exit);
```

**Reply by oleg-shilo** [reply]
Unfortunately, WiX does not have support for showing MSI UI that is not a native MSI UI but a custom managed UI defined by the user. In MSI terminology such UI is called [EmbeddedUI](https://learn.microsoft.com/en-us/windows/win32/msi/using-an-embedded-ui). This is the UI type you are using. 

WiX has had this bug since v3 (https://github.com/wixtoolset/issues/issues/4921) and sadly with WiX4/5 there is still no way to solve this problem directly.
    
Thus your options are rather limited and they are:

- Follow 'WixBootstrapper_MsiEmbeddedUI' sample, which shows a very simple technique of wrapping msi into a self-hosted executable included in the bundle that happily shows the MSI UI regardless of whether it is a native or a managed one.

- Use NSIS bootstrapper instead. Sadly. WiX Bundle seems to be too fragile and inflexible.

- If the only reason for using a bootstrapper is to do a few simple actions (e.g. assess the environment) before product installation, then you can use an incredibly simple bootstrapper/launcher that is a pure custom CLI application. See 'Self-executable_Msi' sample.

**Reply by Lindsay-Mathieson** [reply]
> * Follow 'WixBootstrapper_MsiEmbeddedUI' sample, which shows a very simple technique of wrapping msi into a self-hosted executable included in the bundle that happily shows the MSI UI regardless of whether it is a native or a managed one.

Thanks, that worked well for me.

>See 'Self-executable_Msi' sample.

Interesting, thanks, will assess that, since I just want to run winget

**Reply by oleg-shilo** [reply]
BTW, 'WixBootstrapper_MsiEmbeddedUI' involves building 'Self-executable_Msi' as an authoring step. Thus both samples are very related.
It's just one gives you the opportunity to have some "free-style" routine in the 'Self-executable_Msi' exe and another sample takes that exe into a bundle the traditional way.
